### PR TITLE
4844 on bsc

### PR DESCRIPTION
**Pull request by buddh0**
### Description
BSC will support Cancun EIPs, check out [Tycho hard fork plan](https://forum.bnbchain.org/t/bnb-chain-upgrades-mainnet/936#tycho-wip-8). Other EIPs except EIP-4844 are already supported. But the implementation of EIP-4844 will be the big one and it will be different on BSC, mainly the blob part, as Ethereum store and propagate blobs in its consensus layer, while BSC doesn't have it. BSC need to implement in execution layer.

This PR will implement the EIP-4844 on BSC, there are some docs you may refer:
- [BEP-336: Implement EIP-4844: Shard Blob Transactions](https://github.com/bnb-chain/BEPs/blob/master/BEPs/BEP-336.md)
- [Design: EIP-4844 On BSC](https://forum.bnbchain.org/t/design-eip-4844-on-bsc/2621)
- [FAQ: About EIP-4844 on BSC](https://forum.bnbchain.org/t/faq-about-eip-4844-on-bsc/2620)

### Rationale
Check out the motivation part of BEP-336: https://github.com/bnb-chain/BEPs/blob/master/BEPs/BEP-336.md#4-motivation

### Example
NA

### Changes
- New transaction type(type: 0x3), Blob-Carrying Transaction, its definitin will be same as Ethereum.
- New RPC APIs, mainly to support query of Blobs.
- New blobs related storage in MainDB(level or pebble) and AncientDb.

### CODE REVIEW

File: .github/workflows/build-test.yml [hunk 1]
```diff
@@ -10,6 +10,7 @@ on:
     branches:
       - master
       - develop
+      - blobtx
```
Comment (buddh0) [inline .github/workflows/build-test.yml:13]: remove before merging

**Comment by NathanBSC** [comment]
need add logic to add the blobfee as validator income instead of burning it

File: eth/protocols/eth/protocol.go [hunk 2]
```diff
@@ -215,8 +215,9 @@ type BlockHeadersRLPPacket struct {
 
 // NewBlockPacket is the network packet for the block propagation message.
 type NewBlockPacket struct {
-	Block *types.Block
-	TD    *big.Int
+	Block    *types.Block
+	TD       *big.Int
+	Sidecars []*types.BlobTxSidecar `rlp:"optional"`
```
Comment (NathanBSC) [inline eth/protocols/eth/protocol.go:220]: I prefer    
```
sidecars types.BlobTxSidecars `rlp:"optional"`
```
so can keep same with blockPropagation

File: eth/fetcher/block_fetcher.go [hunk 3]
```diff
@@ -576,7 +576,8 @@ func (f *BlockFetcher) loop() {
 					case res := <-resCh:
 						res.Done <- nil
 						// Ignoring withdrawals here, since the block fetcher is not used post-merge.
-						txs, uncles, _ := res.Res.(*eth.BlockBodiesResponse).Unpack()
+						// todo 4844 is it ok to ignore sidecars here too?
```
Comment (NathanBSC) [inline eth/fetcher/block_fetcher.go:579]: I think It is not ok

File: consensus/parlia/parlia.go [hunk 4]
```diff
@@ -583,23 +583,27 @@ func (p *Parlia) verifyHeader(chain consensus.ChainHeaderReader, header *types.H
...
+		case header.WithdrawalsHash != nil:
+			return fmt.Errorf("invalid WithdrawalsHash, have %#x, expected nil", header.WithdrawalsHash)
 		}
 	} else {
+		switch {
+		case header.ParentBeaconRoot != nil:
+			return fmt.Errorf("invalid parentBeaconRoot, have %#x, expected nil", header.ParentBeaconRoot)
+		// types.EmptyWithdrawalsHash represents a empty value when EIP-4895 enabled,
+		// here, EIP-4895 still be disabled, value expected to be `common.Hash{}` is only to feet the demand of rlp encode/decode
+		case header.WithdrawalsHash == nil || *header.WithdrawalsHash != common.Hash{}:
```
Comment (zzzckck) [inline consensus/parlia/parlia.go:604]: could remove `header.WithdrawalsHash == nil` as WithdrawalsHash will not be nil after Cancun.

File: core/blockchain.go [hunk 5]
```diff
@@ -110,6 +110,7 @@ const (
 	blockCacheLimit     = 256
 	diffLayerCacheLimit = 1024
 	receiptsCacheLimit  = 10000
+	blobsCacheLimit     = 10000
```
Comment (zzzckck) [inline core/blockchain.go:113]: would it be too much? 10000*3*128KB -> 3.6GB.
As it is mainly used by DA check on block import, maybe 1000 could enough

File: core/types/block.go [hunk 6]
```diff
@@ -232,6 +232,9 @@ type Block struct {
 	// inter-peer block relay.
 	ReceivedAt   time.Time
 	ReceivedFrom interface{}
+
+	// blobs provides DA check
+	blobs BlobTxSidecars
```
Comment (zzzckck) [inline core/types/block.go:237]: better to name it as "sidecars", not "blobs"

**Comment by buddh0** [comment]
> need add logic to add the blobfee as validator income instead of burning it

solved

File: cmd/utils/flags.go [hunk 7]
```diff
@@ -1100,6 +1100,13 @@ Please note that --` + MetricsHTTPFlag.Name + ` must be set to start the server.
 		Usage:    "Path for the voteJournal dir in fast finality feature (default = inside the datadir)",
 		Category: flags.FastFinalityCategory,
 	}
+
+	// Blob setting
+	BlobExtraReserveFlag = &cli.Uint64Flag{
+		Name:  "blob.extra-reserve",
+		Usage: "Extra reserve threshold for blob, blob never expires when 0 is set, default 28800",
+		Value: params.DefaultExtraReserveForBlobRequests,
+	}
```
Comment (NathanBSC) [inline cmd/utils/flags.go:1109]: need Category
set to BlockHistoryCategory?

File: consensus/parlia/parlia.go (hunk 4 above)
Comment (NathanBSC) [reply consensus/parlia/parlia.go:604]: > could remove `header.WithdrawalsHash == nil` as WithdrawalsHash will not be nil after Cancun.

if so
a malicious validator generate a bad header, then panic all other validatos

File: core/blockchain.go [hunk 8]
```diff
@@ -359,6 +362,7 @@ func NewBlockChain(db ethdb.Database, cacheConfig *CacheConfig, genesis *Genesis
 		bodyCache:          lru.NewCache[common.Hash, *types.Body](bodyCacheLimit),
 		bodyRLPCache:       lru.NewCache[common.Hash, rlp.RawValue](bodyCacheLimit),
 		receiptsCache:      lru.NewCache[common.Hash, []*types.Receipt](receiptsCacheLimit),
+		sidecarsCache:      lru.NewCache[common.Hash, types.BlobTxSidecars](blobsCacheLimit),
```
Comment (NathanBSC) [inline core/blockchain.go:365]: blobsCacheLimit need rename to sidecarsCacheLimit?

File: eth/downloader/downloader.go [hunk 9]
```diff
@@ -591,6 +594,8 @@ func (d *Downloader) syncWithPeer(p *peerConnection, hash common.Hash, td, ttd *
 	} else if mode == FullSync {
 		fetchers = append(fetchers, func() error { return d.processFullSyncContent(ttd, beaconMode) })
 	}
+	// update the chasing head
+	d.blockchain.UpdateChasingHead(remoteHeader)
```
Comment (NathanBSC) [inline eth/downloader/downloader.go:598]: need compare UpdateChasingHead with remoteHeader firstly?
different peers may have different highest header

File: eth/protocols/eth/handlers.go [hunk 10]
```diff
@@ -349,7 +366,7 @@ func handleBlockBodies(backend Backend, msg Decoder, peer *Peer) error {
 		for i, body := range res.BlockBodiesResponse {
 			txsHashes[i] = types.DeriveSha(types.Transactions(body.Transactions), hasher)
 			uncleHashes[i] = types.CalcUncleHash(body.Uncles)
-			if body.Withdrawals != nil {
+			if len(body.Withdrawals) > 0 {
```
Comment (NathanBSC) [inline eth/protocols/eth/handlers.go:369]: why not need to add sidecars here?

File: consensus/misc/eip4844/eip4844.go [hunk 11]
```diff
@@ -76,8 +76,19 @@ func CalcExcessBlobGas(parentExcessBlobGas uint64, parentBlobGasUsed uint64) uin
...
 // CalcBlobFee calculates the blobfee from the header's excess blob gas field.
-func CalcBlobFee(excessBlobGas uint64) *big.Int {
-	return fakeExponential(minBlobGasPrice, new(big.Int).SetUint64(excessBlobGas), blobGaspriceUpdateFraction)
+// config param aims to support eth's test suit.
+func CalcBlobFee(excessBlobGas uint64, config *params.ChainConfig) *big.Int {
+	dynamicPrice := fakeExponential(minBlobGasPrice, new(big.Int).SetUint64(excessBlobGas), blobGaspriceUpdateFraction)
+	if config == nil || config.Parlia == nil {
+		return dynamicPrice
+	}
+	if dynamicPrice.Cmp(params.MinBlobGasPriceInBSC) < 0 {
```
Comment (buddh0) [inline consensus/misc/eip4844/eip4844.go:85]: I think we can remove the both limits from design, need a talk

File: consensus/parlia/parlia.go (hunk 4 above)
Comment (zzzckck) [reply consensus/parlia/parlia.go:604]: ok, noted

File: core/blockchain.go (hunk 8 above)
Comment (zzzckck) [reply core/blockchain.go:365]: agree

File: core/rawdb/freezer.go [hunk 12]
```diff
@@ -167,6 +181,20 @@ func NewFreezer(datadir string, namespace string, readonly bool, offset uint64,
 	return freezer, nil
 }
 
+// openAdditionTable create table, it will auto create new files when it was first initialized
+func openAdditionTable(datadir, name string, readMeter, writeMeter metrics.Meter, sizeGauge metrics.Gauge, maxTableSize uint32, disableSnappy, readonly bool) (*freezerTable, error) {
```
Comment (zzzckck) [inline core/rawdb/freezer.go:185]: why need this function?

**Review by zzzckck — APPROVED** [review]
LGTM

File: internal/ethapi/api.go [hunk 13]
```diff
@@ -2100,6 +2159,17 @@ func marshalReceipt(receipt *types.Receipt, blockHash common.Hash, blockNumber u
 	return fields
 }
 
+func marshalBlobSidecar(blobSidecar *types.BlobTxSidecar, blockHash common.Hash, blockNumber uint64, tx *types.Transaction, txIndex int) map[string]interface{} {
+	fields := map[string]interface{}{
+		"blockHash":   blockHash,
+		"blockNumber": hexutil.Uint64(blockNumber),
+		"txHash":      tx.Hash(),
+		"txIndex":     hexutil.Uint64(txIndex),
```
Comment (buddh0) [inline internal/ethapi/api.go:2167]: It's better to add "blobHashes" field?

File: consensus/misc/eip4844/eip4844.go (hunk 11 above)
Comment (buddh0) [inline consensus/misc/eip4844/eip4844.go:85]: remove min limit, have confirmed

File: internal/ethapi/api.go (hunk 13 above)
Comment (buddh0) [reply internal/ethapi/api.go:2167]: > It's better to add "blobHashes" field?

 add the blobHashes field if someone really request it  is also ok to me

File: cmd/utils/flags.go (hunk 7 above)
Comment (emailtovamos) [reply cmd/utils/flags.go:1104]: Category for naming or something else?

File: core/rawdb/accessors_chain.go [hunk 14]
```diff
@@ -360,6 +360,20 @@ func ReadHeader(db ethdb.Reader, hash common.Hash, number uint64) *types.Header
 	return header
 }
 
+// ReadHeaderAndRaw retrieves the block header corresponding to the hash.
+func ReadHeaderAndRaw(db ethdb.Reader, hash common.Hash, number uint64) (*types.Header, rlp.RawValue) {
```
Comment (buddh0) [inline core/rawdb/accessors_chain.go:364]: do we really need this?
replace by ReadHeader, and check if the return value is nil, where ReadHeaderAndRaw used

File: core/rawdb/accessors_chain.go [hunk 15]
```diff
@@ -829,12 +1010,33 @@ func writeAncientBlock(op ethdb.AncientWriteOp, block *types.Block, header *type
 	return nil
 }
 
+func writeAncientSidecar(op ethdb.AncientWriteOp, num uint64, blobs types.BlobTxSidecars) error {
```
Comment (buddh0) [inline core/rawdb/accessors_chain.go:1013]: Sidecar --> Sidecars

File: core/rawdb/accessors_chain.go [hunk 16]
```diff
@@ -809,6 +823,173 @@ func WriteAncientBlocks(db ethdb.AncientWriter, blocks []*types.Block, receipts
...
+		log.Crit("Failed to encode block blobs", "err", err)
+	}
+	// Store the flattened receipt slice
+	if err := db.Put(blockBlobSidecarsKey(number, hash), data); err != nil {
+		log.Crit("Failed to store block blobs", "err", err)
+	}
+}
+
+// DeleteBlobSidecars removes all blob data associated with a block hash.
+func DeleteBlobSidecars(db ethdb.KeyValueWriter, hash common.Hash, number uint64) {
```
Comment (buddh0) [inline core/rawdb/accessors_chain.go:987]: not called outside the package
so DeleteBlobSidecars-->deleteBlobSidecars

File: core/rawdb/accessors_chain.go [hunk 17]
```diff
@@ -829,12 +1010,33 @@ func writeAncientBlock(op ethdb.AncientWriteOp, block *types.Block, header *type
 	return nil
 }
 
+func writeAncientSidecar(op ethdb.AncientWriteOp, num uint64, blobs types.BlobTxSidecars) error {
+	if err := op.Append(ChainFreezerBlobSidecarTable, num, blobs); err != nil {
```
Comment (buddh0) [inline core/rawdb/accessors_chain.go:1014]: can we move the logic into func `writeAncientBlock`?
check len(block.Sidecars()) firstly, and by this, we can remove writeAncientBlockWithSidecar
and simplify WriteAncientBlocksAfterCancun

**Comment by NathanBSC** [comment]
mev module should be adapted

File: core/rawdb/chain_freezer.go [hunk 18]
```diff
@@ -301,3 +414,29 @@ func (f *chainFreezer) freezeRange(nfdb *nofreezedb, number, limit uint64) (hash
 
 	return hashes, err
 }
+
+// EmptyBlobAncient check if empty in blob ancient, it is used to init blob ancient
+func EmptyBlobAncient(f *chainFreezer) (bool, error) {
```
Comment (NathanBSC) [inline core/rawdb/chain_freezer.go:419]: not used?

**Comment by buddh0** [comment]
> mev module should be adapted

fixed

**Comment by buddh0** [comment]
> need add logic to add the blobfee as validator income instead of burning it

solved

File: cmd/utils/flags.go (hunk 7 above)
Comment (buddh0) [reply cmd/utils/flags.go:1122]: > Category for naming or something else?

the Category for the helper command to display

**Review by emailtovamos — APPROVED** [review]

**Review by zzzckck — APPROVED** [review]
LGTM👍

**Pull request merged**